File: functions/src/services/FirestoreSessionService.ts
import {
    BaseSessionService,
    CreateSessionRequest,
    GetSessionRequest,
    ListSessionsRequest,
    ListSessionsResponse,
    DeleteSessionRequest,
    AppendEventRequest,
    type Session,
    type Event,
    isFinalResponse,
} from '@google/adk';
import { Firestore, Timestamp } from 'firebase-admin/firestore';
import { summarizeConversation } from './summaryService.js';
import { SUMMARY_EVENT_COUNT_KEY, SUMMARY_STATE_KEY } from '../adk/agents/advisorWorkflowState.js';

const SUMMARY_SKIP_AUTHORS = new Set([
    'signals_research_agent',
    'technical_research_agent',
    'news_research_agent',
    'rag_research_agent',
    'memory_research_agent',
    'search_research_agent',
    'vertex_search_agent',
    'vertex_rag_agent',
]);

export class FirestoreSessionService extends BaseSessionService {
    constructor(private db: Firestore) {
        super();
    }

    async createSession({ appName, userId, state, sessionId }: CreateSessionRequest): Promise<Session> {
        const id = sessionId || `session_${userId}_${Date.now()}`;
        const session: Session = {
            id,
            appName,
            userId,
            state: state || {},
            events: [],
            lastUpdateTime: Date.now(),
        };

        await this.db.collection('sessions').doc(id).set({
            ...session,
            lastUpdateTime: Timestamp.fromMillis(session.lastUpdateTime),
        });

        return session;
    }

    async getSession({ appName, userId, sessionId, config }: GetSessionRequest): Promise<Session | undefined> {
        const doc = await this.db.collection('sessions').doc(sessionId).get();
        if (!doc.exists) return undefined;

        const data = doc.data() as any;
        
        // Basic validation that it belongs to the user and app
        if (data.userId !== userId || data.appName !== appName) {
            return undefined;
        }

        const session: Session = {
            id: data.id,
            appName: data.appName,
            userId: data.userId,
            state: data.state || {},
            events: data.events || [],
            lastUpdateTime: data.lastUpdateTime?.toMillis() || Date.now(),
        };

        // Handle config (numRecentEvents, afterTimestamp)
        if (config) {
            if (config.afterTimestamp) {
                session.events = session.events.filter(e => e.timestamp > config.afterTimestamp!);
            }
            if (config.numRecentEvents) {
                session.events = session.events.slice(-config.numRecentEvents);
            }
        }

        return session;
    }

    async listSessions({ appName, userId }: ListSessionsRequest): Promise<ListSessionsResponse> {
        const snapshot = await this.db.collection('sessions')
            .where('appName', '==', appName)
            .where('userId', '==', userId)
            .orderBy('lastUpdateTime', 'desc')
            .get();

        const sessions = snapshot.docs.map(doc => {
            const data = doc.data();
            return {
                id: data.id,
                appName: data.appName,
                userId: data.userId,
                state: {}, // listSessions usually doesn't need full state/events
                events: [],
                lastUpdateTime: data.lastUpdateTime?.toMillis() || Date.now(),
            } as Session;
        });

        return { sessions };
    }

    async deleteSession({ appName, userId, sessionId }: DeleteSessionRequest): Promise<void> {
        const doc = await this.db.collection('sessions').doc(sessionId).get();
        if (doc.exists) {
            const data = doc.data();
            if (data?.userId === userId && data?.appName === appName) {
                await doc.ref.delete();
            }
        }
    }

    async appendEvent(request: AppendEventRequest): Promise<Event> {
        const { session } = request;
        const event = await super.appendEvent(request);
        
        session.lastUpdateTime = Date.now();

        await this.maybeSummarizeSession(session, event);

        const limitRaw = Number(process.env.SESSION_EVENT_LIMIT);
        const maxEvents = Number.isFinite(limitRaw) ? limitRaw : 50;
        if (maxEvents > 0 && session.events.length > maxEvents) {
            session.events = session.events.slice(-maxEvents);
        }
        
        // Clean events to remove undefined values (Firestore doesn't allow them)
        const cleanEvents = session.events.map(e => JSON.parse(JSON.stringify(e)));
        
        // Update Firestore
        await this.db.collection('sessions').doc(session.id).update({
            events: cleanEvents,
            lastUpdateTime: Timestamp.fromMillis(session.lastUpdateTime),
            state: session.state || {},
        });

        return event;
    }

    private async maybeSummarizeSession(session: Session, event: Event): Promise<void> {
        if (!event.author || event.author === 'user') {
            return;
        }
        if (SUMMARY_SKIP_AUTHORS.has(event.author)) {
            return;
        }
        if (!isFinalResponse(event)) {
            return;
        }

        const triggerRaw = Number(process.env.SESSION_SUMMARY_TRIGGER);
        const keepRaw = Number(process.env.SESSION_SUMMARY_KEEP);
        const cooldownRaw = Number(process.env.SESSION_SUMMARY_COOLDOWN);

        const trigger = Number.isFinite(triggerRaw) ? triggerRaw : 40;
        const keep = Number.isFinite(keepRaw) ? keepRaw : 12;
        const cooldown = Number.isFinite(cooldownRaw) ? cooldownRaw : 20;

        if (trigger <= 0 || keep <= 0 || session.events.length <= trigger) {
            return;
        }

        const lastCountRaw = session.state?.[SUMMARY_EVENT_COUNT_KEY];
        const lastCount = typeof lastCountRaw === 'number' ? lastCountRaw : Number(lastCountRaw ?? 0);
        if (session.events.length < lastCount + cooldown) {
            return;
        }

        const summaryEvents = session.events.slice(0, Math.max(0, session.events.length - keep));
        const existingSummary = typeof session.state?.[SUMMARY_STATE_KEY] === 'string'
            ? session.state[SUMMARY_STATE_KEY]
            : '';

        const summary = await summarizeConversation({
            events: summaryEvents,
            existingSummary,
        });

        if (!summary) {
            return;
        }

        session.state = {
            ...session.state,
            [SUMMARY_STATE_KEY]: summary,
            [SUMMARY_EVENT_COUNT_KEY]: session.events.length,
        };
        session.events = session.events.slice(-keep);
    }

    async updateSession(request: { appName: string; userId: string; sessionId: string; state: any }): Promise<void> {
        await this.db.collection('sessions').doc(request.sessionId).update({
            state: request.state,
            lastUpdateTime: Timestamp.now(),
        });
    }
}

File: functions-python/adk/firestore_session_service.py
"""Firestore-backed session service for ADK."""

from __future__ import annotations

import time
from typing import Any, Optional

import firebase_admin
from firebase_admin import firestore
from google.adk.events import Event
from google.adk.sessions.base_session_service import BaseSessionService, GetSessionConfig, ListSessionsResponse
from google.adk.sessions.session import Session
from . import config
from .genai_client import summarize_conversation

SUMMARY_SKIP_AUTHORS = {
    'signals_research_agent',
    'technical_research_agent',
    'news_research_agent',
    'rag_research_agent',
    'memory_research_agent',
    'search_research_agent',
    'vertex_search_agent',
    'vertex_rag_agent',
}


def _ensure_firebase() -> None:
    if not firebase_admin._apps:
        firebase_admin.initialize_app()


def _serialize_event(event: Event) -> dict[str, Any]:
    return event.model_dump(by_alias=True, mode='json', exclude_none=True)


def _deserialize_event(raw: dict[str, Any]) -> Event:
    return Event.model_validate(raw)


def _get_timestamp_value(value: Any) -> float:
    if value is None:
        return time.time()
    if hasattr(value, 'timestamp'):
        return value.timestamp()
    if isinstance(value, (int, float)):
        return float(value)
    return time.time()


class FirestoreSessionService(BaseSessionService):
    def __init__(self) -> None:
        _ensure_firebase()
        self._db = firestore.client()

    async def create_session(
        self,
        *,
        app_name: str,
        user_id: str,
        state: Optional[dict[str, Any]] = None,
        session_id: Optional[str] = None,
    ) -> Session:
        session_id = session_id or f'session_{user_id}_{int(time.time() * 1000)}'
        session = Session(
            id=session_id,
            app_name=app_name,
            user_id=user_id,
            state=state or {},
            events=[],
            last_update_time=time.time(),
        )

        doc = session.model_dump(by_alias=True, mode='json', exclude_none=True)
        doc['lastUpdateTime'] = firestore.SERVER_TIMESTAMP
        self._db.collection('sessions').document(session_id).set(doc)
        return session

    async def get_session(
        self,
        *,
        app_name: str,
        user_id: str,
        session_id: str,
        config: Optional[GetSessionConfig] = None,
    ) -> Optional[Session]:
        doc_ref = self._db.collection('sessions').document(session_id)
        snapshot = doc_ref.get()
        if not snapshot.exists:
            return None
        data = snapshot.to_dict() or {}
        if data.get('userId') != user_id or data.get('appName') != app_name:
            return None

        raw_events = data.get('events') or []
        events = [_deserialize_event(e) for e in raw_events if isinstance(e, dict)]
        last_update = _get_timestamp_value(data.get('lastUpdateTime'))

        session = Session(
            id=data.get('id', session_id),
            app_name=data.get('appName', app_name),
            user_id=data.get('userId', user_id),
            state=data.get('state') or {},
            events=events,
            last_update_time=last_update,
        )

        if config:
            if config.after_timestamp is not None:
                session.events = [e for e in session.events if e.timestamp > config.after_timestamp]
            if config.num_recent_events:
                session.events = session.events[-config.num_recent_events :]

        return session

    async def list_sessions(self, *, app_name: str, user_id: Optional[str] = None) -> ListSessionsResponse:
        query = self._db.collection('sessions').where('appName', '==', app_name)
        if user_id:
            query = query.where('userId', '==', user_id)
        query = query.order_by('lastUpdateTime', direction=firestore.Query.DESCENDING)

        sessions = []
        for doc in query.stream():
            data = doc.to_dict() or {}
            sessions.append(
                Session(
                    id=data.get('id', doc.id),
                    app_name=data.get('appName', app_name),
                    user_id=data.get('userId', user_id or ''),
                    state={},
                    events=[],
                    last_update_time=_get_timestamp_value(data.get('lastUpdateTime')),
                )
            )
        return ListSessionsResponse(sessions=sessions)

    async def delete_session(self, *, app_name: str, user_id: str, session_id: str) -> None:
        doc_ref = self._db.collection('sessions').document(session_id)
        snapshot = doc_ref.get()
        if not snapshot.exists:
            return
        data = snapshot.to_dict() or {}
        if data.get('appName') == app_name and data.get('userId') == user_id:
            doc_ref.delete()

    async def append_event(self, session: Session, event: Event) -> Event:
        event = await super().append_event(session, event)
        if event.partial:
            return event

        session.last_update_time = time.time()
        await self._maybe_summarize_session(session, event)

        if config.SESSION_EVENT_LIMIT and len(session.events) > config.SESSION_EVENT_LIMIT:
            session.events = session.events[-config.SESSION_EVENT_LIMIT :]

        serialized_events = [_serialize_event(e) for e in session.events]
        self._db.collection('sessions').document(session.id).update({
            'events': serialized_events,
            'lastUpdateTime': firestore.SERVER_TIMESTAMP,
            'state': session.state or {},
        })

        return event

    async def update_session(self, *, app_name: str, user_id: str, session_id: str, state: dict[str, Any]) -> None:
        self._db.collection('sessions').document(session_id).update({
            'state': state,
            'lastUpdateTime': firestore.SERVER_TIMESTAMP,
        })

    async def _maybe_summarize_session(self, session: Session, event: Event) -> None:
        if not event.author or event.author == 'user':
            return
        if event.author in SUMMARY_SKIP_AUTHORS:
            return
        if not event.is_final_response():
            return

        trigger = config.SESSION_SUMMARY_TRIGGER or 40
        keep = config.SESSION_SUMMARY_KEEP or 12
        cooldown = config.SESSION_SUMMARY_COOLDOWN or 20

        if trigger <= 0 or keep <= 0 or len(session.events) <= trigger:
            return

        last_count_raw = session.state.get(config.SUMMARY_EVENT_COUNT_KEY, 0)
        try:
            last_count = int(last_count_raw)
        except (TypeError, ValueError):
            last_count = 0

        if len(session.events) < last_count + cooldown:
            return

        summary_events = session.events[: max(0, len(session.events) - keep)]
        existing_summary = session.state.get(config.SUMMARY_STATE_KEY, '')

        summary = summarize_conversation(summary_events, existing_summary if isinstance(existing_summary, str) else None)
        if not summary:
            return

        session.state = {
            **session.state,
            config.SUMMARY_STATE_KEY: summary,
            config.SUMMARY_EVENT_COUNT_KEY: len(session.events),
        }
        session.events = session.events[-keep:]
